base/operators.jl

```diff
@@ -1149,40 +1149,68 @@ julia> filter(!isletter, str)
 !(f::ComposedFunction{typeof(!)}) = f.inner #allows !!f === f
 
 """
-    Fix1(f, x)
+    Fix{n}(f, x)
+    Fix{:kw}(f, x)
+    Fix(f; kw=x)
 
-A type representing a partially-applied version of the two-argument function
-`f`, with the first argument fixed to the value "x". In other words,
-`Fix1(f, x)` behaves similarly to `y->f(x, y)`.
+A type representing a partially-applied version of a function `f`, with the argument
+"x" fixed at argument `n::Int` or keyword `kw::Symbol`.
+In other words, `Fix{3}(f, x)` behaves similarly to
+`(y1, y2, y3) -> f(y1, y2, x, y3)` for the 4-argument function `f`.
 
-See also [`Fix2`](@ref Base.Fix2).
+You may also use this to fix keyword arguments. For example, `Fix(g; a=2)` behaves
+similarly to `x -> g(x; a=2)` for a function `g` with one argument and one keyword argument.
+You can also write this as `Fix{:a}(g, 2)`.
+
+!!! note
+    When nesting multiple `Fix`, note that the `N` in `Fix{N}` is _relative_ to the current
+    available arguments, rather than an absolute ordering on the target function. For example,
+    `Fix{1}(Fix{2}(f, 4), 4)` fixes the first and second arg, while `Fix{2}(Fix{1}(f, 4), 4)`
+    fixes the first and third arg.
 """
-struct Fix1{F,T} <: Function
+struct Fix{N,F,T} <: Function
     f::F
     x::T
 
-    Fix1(f::F, x) where {F} = new{F,_stable_typeof(x)}(f, x)
-    Fix1(f::Type{F}, x) where {F} = new{Type{F},_stable_typeof(x)}(f, x)
+    function Fix{N}(f::F, x) where {N,F}
+        if N isa Int && N < 1
+            throw(ArgumentError("expected `N` in `Fix{N}` to be integer greater than 0"))
+        elseif !(N isa Union{Int,Symbol})
+            throw(ArgumentError("expected type parameter in `Fix` to be `Int` or `Symbol`, but got type=$(typeof(N))"))
+        end
+        new{N,_stable_typeof(f),_stable_typeof(x)}(f, x)
+    end
+end
+function Fix(f::F; kws...) where {F}
+    length(kws) != 1 && throw(ArgumentError("`Fix` expects exactly one argument or keyword argument"))
+    Fix{only(keys(kws))}(f, only(values(kws)))
+end
+
+function (f::Fix{N})(args::Vararg{Any,M}; kws...) where {N,M}
+    if N isa Symbol
+        N in keys(kws) && throw(ArgumentError("found duplicate keyword argument passed to `Fix`"))
+        f_kws = NamedTuple{(N,)}((f.x,))
+        return f.f(args...; f_kws..., kws...)
+    else # Int
+        M < N-1 && throw(ArgumentError("expected at least $(N-1) arguments to a `Fix` function with `N=$(N)`"))
+        return f.f(args[begin:begin+(N-2)]..., f.x, args[begin+(N-1):end]...; kws...)
```

Comment:
Neither one seems to strengthen the constprop without special casing.

But note I'm not sure that one constprop test is actually an issue; it is a fairly difficult constprop task that the compiler doesn't perform with any `Vararg` in the way (even if specialisation is forced), see https://github.com/JuliaLang/julia/issues/55042. I think we should just stick with the special casing to prevent any performance regressions on Fix1/Fix2, and maybe constprop will continue to be refined in the future to improve the general version.

The first snippet would result in worse performance because `first(::Tuple, ::Int)` returns a vector, however, the second snippet you sent seems like it might be nice in general because `getindex(::Tuple, ::UnitRange)` seems to only specialize up to length 10:

https://github.com/JuliaLang/julia/blob/aa0758594f0244f33fc4c793c344b1ce227a56eb/base/range.jl#L436-L451

What do you think @LilithHafner, should `ntuple` be called explicitly here to construct the args?

Reply:
This doesn't improve the constprop either, sadly. Still need the special cases even with `ntuple(i->..., Val(M+1))`.

Keep in mind it's not an inference issue – `Fix` itself is perfectly type stable with any of these options. It's just that the compiler is picky about when it does constprop, and simply having a `Vararg` is enough to prevent it in that one test (see #55042)

Reply:
Did you have some examples of this? If you do then we should try to figure out why it’s happening and fix it. But if it’s only that one constprop test then I’m not sure the general advice applies any more than it would to creating aliases in general.

Reply:
Oh I see what you mean. Well, this is a more general observation about constant propagation and inlining so I think is better suited for the performance page rather than the docstring of `Fix`. What's your call @LilithHafner?

Reply:
The orders of magnitude slowdown is not introduced by this PR:

```julia
julia> @b map(x -> x.a, tbl)
322.501 μs (10 allocs: 7.630 MiB)

julia> @b map(x -> getproperty(x, :a), tbl)
319.127 μs (10 allocs: 7.630 MiB)

julia> @b map(Base.Fix2(getproperty, :a), tbl)
21.499 ms (1999499 allocs: 53.398 MiB, 1.72% gc time)

julia> VERSION
v"1.10.4"
```

I don't see anything merge-blocking in this discussion, though there are certainly valuable areas of improvement that could be moved to issues and/or performance tips